lib/spack/spack/spec.py

```diff
@@ -4871,6 +4872,46 @@ def reconstruct_virtuals_on_edges(spec):
             edge.update_virtuals([vspec])
 
 
+def format_path(spec, format_string):
+    """Given a `format_string` that is intended as a path, generate a string
+    like from `Spec.format`, but eliminate extra path separators introduced by
+    formatting of Spec properties.
+
+    Path separators explicitly added to the string are preserved, so for example
+    "{name}/{version}" would generate a directory based on the Spec's name, and
+    a subdirectory based on its version; this function guarantees though that
+    the resulting string would only have two directories (i.e. that if under
+    normal circumstances that `str(Spec.version)` would contain a path
+    separator, it would not in this case).
+
+    Note that POSIX or Windows-style paths can be supplied on either POSIX or
+    Windows, and both "/" and "\" are treated as path separators (as a side
+    effect, "invalid" paths which mix these will be converted to OS-appropriate
+    paths with consistent separators).
+    """
+    # A root on Linux ("/"), or part of a Windows device namespace ("\")
+    any_sep = r"[/\\]"
+    # A drive on Windows (e.g. "C:\\")
+    drive = r"[a-zA-Z]:[/\\]"
+    root_pattern = rf"^({any_sep}|{drive})"
```

Comment:
> Can't you simplify to `os.path.isabs`

I want to avoid calling `os.path` functions on strings which may not be valid paths because it is unclear to me from reading the `os.path` documentation ([here](https://docs.python.org/3/glossary.html#term-path-like-object) and [here](https://docs.python.org/3/library/os.path.html)) whether the strings you pass to these functions should be valid paths. Also note that even if forbidding `\`, `c:/` is a valid root when specifying Windows paths in POSIX format (so without `isabs`, checking for `^/` is not sufficient).

> Or better, I wouldn't mind if projections were always in posix format, since \ is a special character in yaml, and doesn't seem to work how you'd expect, leading only to confusion.

so far in Spack this has been considered tractable, and is handled with escaping, e.g. in `share/spack//qa/configuration/windows_config.yaml`. We could constrain the yaml to be POSIX, but should likely enforce that at the YAML-parsing level (or at least at config reading time, since I don't know if enforcing this constraint with the schema would lead to reasonable error messages). 

Furthermore, if we forbid it here I'd like to be consistent about forbidding it in other spots in the configs, which IMO exceeds the scope of this PR.

> (Minor: `root_pattern = rf"{any_sep}|{drive}"` is sufficient)

I am aware of the difference between match and search, but I still like to use `^` at the beginning of the regular expression string when calling `re.match` (I consider it distracting to subsume regular expression constructs into library function names).

Reply:
I don't find the example `configuration/windows_config.yaml` convincing giving that it literally has 3 unique ways to specify 3 paths:

```yaml
config:
  install_tree:
    # no quote, single \
    root: $spack\opt\spack
    projections:
      # quote: escaped \
      all: '{architecture}\\{compiler.name}-{compiler.version}\\{name}-{version}-{hash}'
  build_stage:
    # posix?!
    - ~/.spack/stage
 ```

Just allowing `/` sounds much less error prone.

Reply:
One thing I forgot:

```
>>> os.path.isabs("c:/") # not on windows
False
```

If we see `c:/` on Linux (e.g. because a `spack.yaml` was moved between systems), I want to alert the user to what is likely an unintended behavior.

> since January 1992.

It doesn't matter to me how long the underlying code has been the same: that is not how to decide what the API of a function is.

>> (so without isabs, checking for ^/ is not sufficient).

> It is sufficient because you're mapping : and \ to _

Doing this, `c:/` would be changed to `c_/`, which means we've silently taken what the user probably intends (a path rooted at `c:/`) and mangled it into something else.

> Another issue: why would a function called format_path throw when it gets an absolute path?

To allow absolute paths, I would need to properly handle the following cases:

1. Windows device namespaces (which start with "\\")
2. Windows absolute paths with drive letters ("c:/")
3. Linux absolute paths

Which have the following considerations

* In case [2], I must not change "c:/" to "c_/"
* In case [3], I need to be careful about how I `split` (note that `\\root\subdir` is not the same as `\root\subdir` and likewise for `//root/subdir`/`/root/subdir`.

Since nothing appears to do need that, I don't want to complicate the function with these concerns (and the function currently contains a comment which makes reference to these issues).

Reply:
This is absolute psuedocode, but wouldn't this work to preserve absolute files on windows?

1. Check for `^[a-z]:[/\\]` and remove it if present, save as `drive`
2. Check for `^(//|\\\\)` and remove it if present, save as `device`
3. format the rest of the string as appropriate, save as `result`
4. return `drive + device + result` (pretty sure either `drive` or `device` will be `""`, but that's fine)

Reply:
That covers most of it but also

* I'd want to be careful about parsing `c:///` (make sure to strip leading slashes after pulling drive)
* I'd want to be careful about [4], joining the root properly (as Harmen points out, `os.path.join` can run into issues, and while experimenting with Windows roots and `pathlib` I've gone awry - e.g. I have to make sure that `C:/` is the first element and not `C:`, else it gets a malformed path)

Overall it seems like extra logic that isn't needed by anyone who calls this function (it's a good point that it isn't a *lot* of extra logic though).